src/hydro/dual_energy.h

```diff
@@ -0,0 +1,277 @@
+ *     implementations.
+ */
+
+#ifndef DUAL_ENERGY_H
+#define DUAL_ENERGY_H
+
+#include "../global/global.h"
+#include "../grid/grid_enum.h"
+#include "../hydro/hydro_cuda.h"
+#include "../utils/basic_structs.h"
+
+namespace dual_energy
+{
+
+/*! Overwrites the total energy with the sum of the internal energy and the kinetic energy
+*
+* \tparam NDim The number of dimensions
+*
+# \param[in,out] dev_conserved Pointer to the conserved quantities on the device
+* \param[in] grid_shape Specifies the shape of the grid (including ghost zones). Values along unused dimensions
+* aren't used.
+* \param[in] n_ghost The number of ghost zones (along each used dimension)
+* \param[in] gamma The adiabatic index
+* \param[in] n_fields The number of fields used in the current simulation.
+*
+* \note
+* The scenario where E > 0 and E doesn't exceed the kinetic energy (i.e. U_total <= 0), comes up with some frequency
+* when we include particle feedback.
+*   * in that scenario, we explicitly use the value held by the U_advected field
+*   * in the separate `Sync_Energies_3D` kernel, we then override the E field with KE + U_advected
+*   * one might argue we should actually override the E field with KE + U_advected before this kernel
+*     (we leave that for future consideration)
+*   * regardless, it is **REALLY IMPORTANT** that the E field is **NOT** modified in this kernel
+*     (modifying it will produce race conditions!!!)
+*/
+template <int NDim>
+__global__ void Select_Internal_Energy(Real *dev_conserved, hydro_utilities::VectorXYZ<int> grid_shape, int n_ghost,
+                                       int n_fields)
+{
+  static_assert((NDim == 1) || (NDim == 2) || (NDim == 3), "NDim must be 1, 2, or 3");
+
+#ifndef DE
+  printf("WARNING: this function isn't usable since Cholla wasn't compiled with Dual Energy Formalism!\n");
+
+#else
+  int id, n_cells;
+  int neighbor_ids[NDim * 2];
+  bool is_real_cell;
+
+  // Here we branch and determine some basic details based on the number of dimensions
+  // -> there are much more clever ways to do all of this more concisely
+  // -> for the uninitiated, the conditions of `constexpr if` statements (within templates) are evaluated
+  //    at compile time. In other words, there is no branching
+  if constexpr (NDim == 1) {
+    n_cells = grid_shape.x();
+    id      = threadIdx.x + blockIdx.x * blockDim.x;
+
+    int nx = grid_shape.x();
+
+    int xid = id;
+
+    int imo         = max(xid - 1, n_ghost);
+    neighbor_ids[0] = imo;
+    int ipo         = min(xid + 1, nx - n_ghost - 1);
+    neighbor_ids[1] = ipo;
+
+    is_real_cell = (xid > n_ghost - 1 && xid < nx - n_ghost);
+
+  } else if constexpr (NDim == 2) {
+    n_cells = grid_shape.x() * grid_shape.y();
+
+    // we maintain the use of blockId for backwards compatability, but I'm not sure it does anything
+    // (i.e. can we just replace blockId with blockIdx.x as in the other 2 branches)?
```

Comment:
At one point in the distant past, we needed to use 2D grids to run really large 2D problems, because of CUDA limits on the 1D grid dimensions. I believe this hasn't been an issue for a long time, but you can see it reflected in the name 'dim2dgrid' for the grid dimensions in e.g. the VL_2D_cuda.cu functions, versus 'dim1dgrid' for the 3D version, even though then we set up a a 1D grid. In principle as long as we continue to only use 1D grids, replacing this with blockIdx.x would be fine, since blockIdx.y will always be 0 for a 1D grid, but that's why it's like this (and the distinction lives on in the other 2D versions of hydro functions).